src/main/java/roomescape/controller/ReservationController.java

```diff
@@ -0,0 +1,63 @@
+package roomescape.controller;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import roomescape.domain.Reservation;
+import roomescape.dto.ReservationRequest;
+import roomescape.dto.ReservationResponse;
+
+@Controller
+@RequestMapping("/reservations")
+public class ReservationController {
+
+    private final List<Reservation> reservations = new ArrayList<>();
+    private final AtomicLong index = new AtomicLong(0);
```

Comment:
AtomicLong의 경우 생성자 파라미터로 아무 값도 넣지 않으면 default 값으로 0을 주입하는 것으로 알고있어요. 😃
물론 이 케이스는 명시적으로 넣어주는 것도 나쁘지 않다고 생각해서, 혹시 모르실까봐 언급드려봤어요. 😉 

```suggestion
    private final AtomicLong index = new AtomicLong();
```

Reply:
추가적으로 그냥 Long을 쓸 수도 있을 것 같은데, AtomicLong을 쓴 이유가 있을까요? 😃 
(혹시 호떡의 선택이 아닐지라도..! 이참에 알아가보면 좋을 것 같아요)

Reply:
AtomicLong과 같은 장점을 `List<Reservation>`의 경우에도 비슷하게 개선해보면 어떨까요?

Reply:
> 그냥 Long을 쓸 수도 있을 것 같은데, AtomicLong을 쓴 이유가 있을까요?

AtomicLong을 사용하면 CAS 알고리즘을 통해 동시성 문제를 해결할 수 있어요.
반면 Long을 사용했을 때 여러 스레드가 index의 값을 read하고 write하는 과정에서 동시성 문제가 발생할 수 있어요. 애플리케이션 레벨의 코드만 봤을 때는 index++가 원자 연산으로 보이지만 바이트 코드만 까봐도 load, add, store 3단계로 이루어지는 것을 확인할 수 있고, 이 때 여러 스레드가 store 하기 전에 메모리의 값을 load 한다면 동시성 문제가 발생할 수 있어요.

물론, 이를 synchronized 블럭으로 감싸 임계구역을 설정하여 해결할 수도 있지만 AtomicLong의 CAS연산은 하드웨어 레벨에서 임계구역이 아닌 값 자체에 대한 get compare write 연산을 원자적으로 수행하면서 직접 값의 변경 여부를 검사하며 동시성 문제를 해결해요. 즉, 락을 대기하며 스레드가 블로킹되고, 컨텍스트 스위칭이 발생하는 오버헤드를 완전히 없앨 수 있으므로 CAS를 이용한 동시성 제어 방식이 훨씬 빠른 방법입니다.

> AtomicLong과 같은 장점을 `List<Reservation>`의 경우에도 비슷하게 개선해보면 어떨까요?

List라는 힙의 공유 자원에 대해서도 동시성 문제가 발생할 수 있다는 점을 생각하지 못했네요 .. 😂
아래와 같이 수정하였습니다.
```
private final List<Reservation> reservations = Collections.synchronizedList(new ArrayList<>());
```